linkerd/app/inbound/src/accept.rs

```diff
@@ -0,0 +1,175 @@
+use crate::{AllowPolicy, Inbound};
+use linkerd_app_core::{
+    io, svc,
+    transport::addrs::{ClientAddr, OrigDstAddr, Remote},
+    Error,
+};
+use std::fmt::Debug;
+use tracing::info_span;
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Accept {
+    client_addr: Remote<ClientAddr>,
+    orig_dst_addr: OrigDstAddr,
+    policy: AllowPolicy,
+}
+
+// === impl Inbound ===
+
+impl<N> Inbound<N> {
+    /// Builds a stack that accepts connections. Connections to the proxy port are diverted to the
+    /// 'direct' stack; otherwise connections are associated with a policy and passed to the inner
+    /// stack.
+    pub fn push_accept<T, I, NSvc, D, DSvc>(
+        self,
+        proxy_port: u16,
+        direct: D,
+    ) -> Inbound<svc::BoxNewTcp<T, I>>
+    where
+        T: svc::Param<Remote<ClientAddr>> + svc::Param<OrigDstAddr>,
+        T: Clone + Send + 'static,
+        I: io::AsyncRead + io::AsyncWrite + io::Peek + io::PeerAddr,
+        I: Debug + Send + Sync + Unpin + 'static,
+        N: svc::NewService<Accept, Service = NSvc> + Clone + Send + Sync + Unpin + 'static,
+        NSvc: svc::Service<I, Response = ()>,
+        NSvc: Send + Unpin + 'static,
+        NSvc::Error: Into<Error>,
+        NSvc::Future: Send,
+        D: svc::NewService<T, Service = DSvc> + Clone + Send + Sync + Unpin + 'static,
+        DSvc: svc::Service<I, Response = ()> + Send + 'static,
+        DSvc::Error: Into<Error>,
+        DSvc::Future: Send,
+    {
+        self.map_stack(|cfg, rt, accept| {
+            let port_policies = cfg.port_policies.clone();
+            accept
+                .push_switch(
+                    move |t: T| -> Result<_, Error> {
+                        let OrigDstAddr(addr) = t.param();
+                        if addr.port() == proxy_port {
+                            return Ok(svc::Either::B(t));
+                        }
+                        let policy = port_policies.check_allowed(addr.port())?;
+                        Ok(svc::Either::A(Accept {
+                            client_addr: t.param(),
+                            orig_dst_addr: t.param(),
+                            policy,
+                        }))
+                    },
```

Comment:
before we had a `prevent_loop` module to implement the first 4 lines of this predicate. now this is in-lined and we additionally check policies to ensure that the port isn't denied.